File: pulsar-skeleton/src/main/java/ai/platon/pulsar/crawl/fetch/indexer/JITIndexer.java
package ai.platon.pulsar.crawl.fetch.indexer;

import ai.platon.pulsar.common.StringUtil;
import ai.platon.pulsar.common.Urls;
import ai.platon.pulsar.common.config.ImmutableConfig;
import ai.platon.pulsar.common.config.Params;
import ai.platon.pulsar.common.config.ReloadableParameterized;
import ai.platon.pulsar.crawl.fetch.FetchMonitor;
import ai.platon.pulsar.crawl.fetch.FetchTask;
import ai.platon.pulsar.crawl.index.IndexDocument;
import ai.platon.pulsar.crawl.index.IndexWriters;
import ai.platon.pulsar.crawl.index.IndexingFilters;
import ai.platon.pulsar.crawl.scoring.ScoringFilters;
import ai.platon.pulsar.persist.ParseStatus;
import ai.platon.pulsar.persist.WebPage;
import ai.platon.pulsar.persist.metadata.ParseStatusCodes;
import com.google.common.collect.Lists;
import com.google.common.collect.Queues;
import org.slf4j.Logger;

import java.time.Instant;
import java.util.LinkedList;
import java.util.Set;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ConcurrentSkipListSet;
import java.util.concurrent.atomic.AtomicInteger;

import static ai.platon.pulsar.common.PulsarParams.DOC_FIELD_TEXT_CONTENT;
import static ai.platon.pulsar.common.config.CapabilityTypes.INDEX_JIT;

/**
 * Created by vincent on 16-8-23.
 * Copyright @ 2013-2016 Platon AI. All rights reserved
 */
public class JITIndexer implements ReloadableParameterized, AutoCloseable {

  public static final Logger LOG = FetchMonitor.LOG;

  private static AtomicInteger instanceSequence = new AtomicInteger(0);

  private final int id;
  private ImmutableConfig conf;
  private boolean indexJIT;
  private final IndexWriters indexWriters;
  private final IndexingFilters indexingFilters;
  private final ScoringFilters scoringFilters;
  private IndexDocument.Builder indexDocumentBuilder;

  private int batchSize = 2000;
  private int indexThreadCount;
  private int minTextLenght;
  // All object inside a process shares the same counters
  private AtomicInteger indexedPages = new AtomicInteger(0);
  private AtomicInteger ignoredPages = new AtomicInteger(0);

  private final Set<IndexThread> activeIndexThreads = new ConcurrentSkipListSet<>();
  private final BlockingQueue<FetchTask> indexTasks = Queues.newLinkedBlockingQueue(batchSize);

  private final LinkedList<IndexThread> indexThreads = Lists.newLinkedList();

  public JITIndexer(ScoringFilters scoringFilters, IndexingFilters indexingFilters, IndexWriters indexWriters, ImmutableConfig conf) {
    this.id = instanceSequence.incrementAndGet();
    this.scoringFilters = scoringFilters;
    this.indexingFilters = indexingFilters;
    this.indexWriters = indexWriters;

    reload(conf);
  }

  @Override
  public ImmutableConfig getConf() {
    return conf;
  }

  @Override
  public Params getParams() {
    return Params.of(
        "batchSize", batchSize,
        "indexThreadCount", indexThreadCount,
        "minTextLenght", minTextLenght
    );
  }

  @Override
  public void reload(ImmutableConfig conf) {
    this.conf = conf;

    this.indexJIT = conf.getBoolean(INDEX_JIT, false);
    this.batchSize = conf.getInt("index.index.batch.size", this.batchSize);
    this.indexThreadCount = conf.getInt("index.index.thread.count", 1);
    this.minTextLenght = conf.getInt("index.minimal.text.length", 300);

    if (indexJIT) {
      this.indexDocumentBuilder = new IndexDocument.Builder(conf).with(indexingFilters).with(scoringFilters);
      this.indexWriters.open();
    }
  }

  public int id() { return id; }

  void registerFetchThread(IndexThread indexThread) {
    activeIndexThreads.add(indexThread);
  }

  void unregisterFetchThread(IndexThread indexThread) {
    activeIndexThreads.remove(indexThread);
  }

  public int getIndexThreadCount() {
    return indexThreadCount;
  }

  public int getIndexedPages() {
    return indexedPages.get();
  }

  public int getIngoredPages() {
    return ignoredPages.get();
  }

  /**
   * Add fetch item to index indexTasks
   * Thread safe
   */
  public void produce(FetchTask fetchTask) {
    if (!indexJIT) {
      return;
    }

    WebPage page = fetchTask.getPage();
    if (page == null) {
      LOG.warn("Invalid FetchTask to index, ignore it");
      return;
    }

    if (!shouldProduce(page)) {
      return;
    }

    indexTasks.add(fetchTask);
  }

  /**
   * Thread safe
   */
  public FetchTask consume() {
    return indexTasks.poll();
  }

  @Override
  public void close() {
    LOG.info("[Destruction] Closing JITIndexer #" + id + " ...");

    indexThreads.forEach(IndexThread::exitAndJoin);

    try {
      FetchTask fetchTask = consume();
      while (fetchTask != null) {
        index(fetchTask);
        fetchTask = consume();
      }
    } catch (Throwable e) {
      LOG.error(e.toString());
    }

    LOG.info("There are " + ignoredPages + " not indexed short pages out of total " + indexedPages + " pages");
  }

  /**
   * Thread safe
   */
  public void index(FetchTask fetchTask) {
    if (!indexJIT) {
      return;
    }

    try {
      if (fetchTask == null) {
        LOG.error("Failed to index, null fetchTask");
        return;
      }

      String url = fetchTask.getUrl();
      String reverseUrl = Urls.reverseUrl(url);
      WebPage page = fetchTask.getPage();

      IndexDocument doc = indexDocumentBuilder.build(reverseUrl, page);
      if (shouldIndex(doc)) {
        synchronized (indexWriters) {
          indexWriters.write(doc);
          page.putIndexTimeHistory(Instant.now());
        }
        indexedPages.incrementAndGet();
      } // if
    } catch (Throwable e) {
      LOG.error("Failed to index a page " + StringUtil.stringifyException(e));
    }
  }

  private boolean shouldIndex(IndexDocument doc) {
    if (doc == null) {
      return false;
    }

    String textContent = doc.getFieldValueAsString(DOC_FIELD_TEXT_CONTENT);
    if (textContent == null || textContent.length() < minTextLenght) {
      ignoredPages.incrementAndGet();
      LOG.warn("Invalid text content to index, url : " + doc.getUrl());
      return false;
    }

    return true;
  }

  private boolean shouldProduce(WebPage page) {
    if (page.isSeed()) {
      return false;
    }

    ParseStatus status = page.getParseStatus();

    if (status == null || !status.isSuccess() || status.getMajorCode() == ParseStatusCodes.SUCCESS_REDIRECT) {
      return false;
    }

    if (page.getContentText().length() < minTextLenght) {
      ignoredPages.incrementAndGet();
      return false;
    }

    return true;
  }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/crawl/fetch/indexer/JITIndexer.kt
package ai.platon.pulsar.crawl.fetch.indexer

import ai.platon.pulsar.common.NetUtil
import ai.platon.pulsar.common.PulsarParams.DOC_FIELD_TEXT_CONTENT
import ai.platon.pulsar.common.config.*
import ai.platon.pulsar.common.config.CapabilityTypes.INDEXER_JIT
import ai.platon.pulsar.common.stringify
import ai.platon.pulsar.common.urls.UrlUtils
import ai.platon.pulsar.crawl.common.JobInitialized
import ai.platon.pulsar.crawl.fetch.batch.JobFetchTask
import ai.platon.pulsar.crawl.index.IndexDocument
import ai.platon.pulsar.crawl.index.IndexWriters
import ai.platon.pulsar.crawl.index.IndexingFilters
import ai.platon.pulsar.crawl.scoring.ScoringFilters
import ai.platon.pulsar.persist.WebPage
import ai.platon.pulsar.persist.metadata.ParseStatusCodes
import com.google.common.collect.Queues
import org.slf4j.LoggerFactory
import java.time.Instant
import java.util.*
import java.util.concurrent.ConcurrentSkipListSet
import java.util.concurrent.atomic.AtomicInteger

/**
 * Created by vincent on 16-8-23.
 * Copyright @ 2013-2016 Platon AI. All rights reserved
 */
class JITIndexer(
    private val scoringFilters: ScoringFilters,
    private val indexingFilters: IndexingFilters,
    private val indexWriters: IndexWriters,
    val conf: ImmutableConfig
) : Parameterized, JobInitialized, AutoCloseable {

    private val id: Int = instanceSequence.incrementAndGet()

    var isEnabled: Boolean = false
        private set

    /**
     * Index server
     */
    val indexServerHost = conf.get(CapabilityTypes.INDEXER_HOSTNAME, AppConstants.DEFAULT_INDEX_SERVER_HOSTNAME)
    val indexServerPort = conf.getInt(CapabilityTypes.INDEXER_PORT, AppConstants.DEFAULT_INDEX_SERVER_PORT)

    private var batchSize: Int = conf.getInt("index.index.batch.size", 2000)
    var indexThreadCount: Int = conf.getInt("index.index.thread.count", 1)
    private var minTextLength: Int = conf.getInt("index.minimal.text.length", 300)
    // All object inside a process shares the same counters
    private val indexedPages = AtomicInteger(0)
    private val ignoredPages = AtomicInteger(0)

    private val indexThreads = mutableListOf<IndexThread>()
    private val activeIndexThreads = ConcurrentSkipListSet<IndexThread>()
    private lateinit var indexTasks: Queue<JobFetchTask>
    private lateinit var indexDocumentBuilder: IndexDocument.Builder

    val indexedPageCount: Int get() = indexedPages.get()
    val ignoredPageCount: Int get() = ignoredPages.get()

    override fun setup(jobConf: ImmutableConfig) {
        isEnabled = jobConf.getBoolean(INDEXER_JIT, false)
        if (isEnabled) {
            indexTasks = Queues.newLinkedBlockingQueue<JobFetchTask>(batchSize)

            this.indexDocumentBuilder = IndexDocument.Builder(conf).with(indexingFilters).with(scoringFilters)
            this.indexWriters.open()
        }
    }

    override fun getParams(): Params {
        return Params.of(
                "isEnabled", isEnabled,
                "batchSize", batchSize,
                "indexThreadCount", indexThreadCount,
                "minTextLength", minTextLength
        )
    }

    fun isIndexServerAvailable(): Boolean {
        return NetUtil.testHttpNetwork(indexServerHost, indexServerPort)
    }

    internal fun registerFetchThread(indexThread: IndexThread) {
        activeIndexThreads.add(indexThread)
    }

    internal fun unregisterFetchThread(indexThread: IndexThread) {
        activeIndexThreads.remove(indexThread)
    }

    /**
     * Add fetch item to index indexTasks
     * Thread safe
     */
    fun produce(fetchTask: JobFetchTask) {
        if (!isEnabled) {
            return
        }

        val page = fetchTask.page

        if (!shouldProduce(page)) {
            return
        }

        indexTasks.add(fetchTask)
    }

    /**
     * Thread safe
     */
    fun consume(): JobFetchTask? {
        return indexTasks.poll()
    }

    override fun close() {
        if (!isEnabled) {
            return
        }

        LOG.info("[Destruction] Closing JITIndexer #$id ...")

        indexThreads.forEach { it.exitAndJoin() }

        try {
            var fetchTask = consume()
            while (fetchTask != null) {
                index(fetchTask)
                fetchTask = consume()
            }
        } catch (e: Throwable) {
            LOG.error(e.toString())
        }

        LOG.info("There are $ignoredPageCount not indexed short pages out of total $indexedPageCount pages")
    }

    /**
     * Thread safe
     */
    fun index(fetchTask: JobFetchTask?) {
        if (!isEnabled) {
            return
        }

        try {
            if (fetchTask == null) {
                LOG.error("Failed to index, null fetchTask")
                return
            }

            val url = fetchTask.urlString
            val reverseUrl = UrlUtils.reverseUrl(url)
            val page = fetchTask.page

            val doc = indexDocumentBuilder.build(reverseUrl, page)
            if (shouldIndex(doc)) {
                synchronized(indexWriters) {
                    indexWriters.write(doc)
                    page.putIndexTimeHistory(Instant.now())
                }
                indexedPages.incrementAndGet()
            } // if
        } catch (e: Throwable) {
            LOG.error("Failed to index a page " + e.stringify())
        }

    }

    private fun shouldIndex(doc: IndexDocument?): Boolean {
        if (doc == null) {
            return false
        }

        val textContent = doc.getFieldValueAsString(DOC_FIELD_TEXT_CONTENT)
        if (textContent == null || textContent.length < minTextLength) {
            ignoredPages.incrementAndGet()
            LOG.warn("Invalid text content to index, url : " + doc.url)
            return false
        }

        return true
    }

    private fun shouldProduce(page: WebPage): Boolean {
        if (page.isSeed) {
            return false
        }

        val status = page.parseStatus

        if (!status.isSuccess || status.majorCode.toInt() == ParseStatusCodes.SUCCESS_REDIRECT) {
            return false
        }

        if (page.contentText.length < minTextLength) {
            ignoredPages.incrementAndGet()
            return false
        }

        return true
    }

    companion object {
        val LOG = LoggerFactory.getLogger(JITIndexer::class.java)
        private val instanceSequence = AtomicInteger(0)
    }
}
